File: docs/source/src/python/user-guide/transformations/joins.py
# --8<-- [start:prep-data]
import pathlib
import requests


DATA = [
    (
        "https://raw.githubusercontent.com/pola-rs/polars-static/refs/heads/master/data/monopoly_props_groups.csv",
        "docs/assets/data/monopoly_props_groups.csv",
    ),
    (
        "https://raw.githubusercontent.com/pola-rs/polars-static/refs/heads/master/data/monopoly_props_prices.csv",
        "docs/assets/data/monopoly_props_prices.csv",
    ),
]


for url, dest in DATA:
    if pathlib.Path(dest).exists():
        continue
    with open(dest, "wb") as f:
        f.write(requests.get(url, timeout=10).content)
# --8<-- [end:prep-data]

# --8<-- [start:props_groups]
import polars as pl

props_groups = pl.read_csv("docs/assets/data/monopoly_props_groups.csv").head(5)
print(props_groups)
# --8<-- [end:props_groups]

# --8<-- [start:props_prices]
props_prices = pl.read_csv("docs/assets/data/monopoly_props_prices.csv").head(5)
print(props_prices)
# --8<-- [end:props_prices]

# --8<-- [start:equi-join]
result = props_groups.join(props_prices, on="property_name")
print(result)
# --8<-- [end:equi-join]

# --8<-- [start:props_groups2]
props_groups2 = props_groups.with_columns(
    pl.col("property_name").str.to_lowercase(),
)
print(props_groups2)
# --8<-- [end:props_groups2]

# --8<-- [start:props_prices2]
props_prices2 = props_prices.select(
    pl.col("property_name").alias("name"), pl.col("cost")
)
print(props_prices2)
# --8<-- [end:props_prices2]

# --8<-- [start:join-key-expression]
result = props_groups2.join(
    props_prices2,
    left_on="property_name",
    right_on=pl.col("name").str.to_lowercase(),
)
print(result)
# --8<-- [end:join-key-expression]

# --8<-- [start:inner-join]
result = props_groups.join(props_prices, on="property_name", how="inner")
print(result)
# --8<-- [end:inner-join]

# --8<-- [start:left-join]
result = props_groups.join(props_prices, on="property_name", how="left")
print(result)
# --8<-- [end:left-join]

# --8<-- [start:right-join]
result = props_groups.join(props_prices, on="property_name", how="right")
print(result)
# --8<-- [end:right-join]

# --8<-- [start:left-right-join-equals]
print(
    result.equals(
        props_prices.join(
            props_groups,
            on="property_name",
            how="left",
            # Reorder the columns to match the order from above.
        ).select(pl.col("group"), pl.col("property_name"), pl.col("cost"))
    )
)
# --8<-- [end:left-right-join-equals]

# --8<-- [start:full-join]
result = props_groups.join(props_prices, on="property_name", how="full")
print(result)
# --8<-- [end:full-join]

# --8<-- [start:full-join-coalesce]
result = props_groups.join(
    props_prices,
    on="property_name",
    how="full",
    coalesce=True,
)
print(result)
# --8<-- [end:full-join-coalesce]

# --8<-- [start:semi-join]
result = props_groups.join(props_prices, on="property_name", how="semi")
print(result)
# --8<-- [end:semi-join]

# --8<-- [start:anti-join]
result = props_groups.join(props_prices, on="property_name", how="anti")
print(result)
# --8<-- [end:anti-join]

# --8<-- [start:players]
players = pl.DataFrame(
    {
        "name": ["Alice", "Bob"],
        "cash": [78, 135],
    }
)
print(players)
# --8<-- [end:players]

# --8<-- [start:non-equi]
result = players.join_where(props_prices, pl.col("cash") > pl.col("cost"))
print(result)
# --8<-- [end:non-equi]

# --8<-- [start:df_trades]
from datetime import datetime

df_trades = pl.DataFrame(
    {
        "time": [
            datetime(2020, 1, 1, 9, 1, 0),
            datetime(2020, 1, 1, 9, 1, 0),
            datetime(2020, 1, 1, 9, 3, 0),
            datetime(2020, 1, 1, 9, 6, 0),
        ],
        "stock": ["A", "B", "B", "C"],
        "trade": [101, 299, 301, 500],
    }
)
print(df_trades)
# --8<-- [end:df_trades]

# --8<-- [start:df_quotes]
df_quotes = pl.DataFrame(
    {
        "time": [
            datetime(2020, 1, 1, 9, 0, 0),
            datetime(2020, 1, 1, 9, 2, 0),
            datetime(2020, 1, 1, 9, 4, 0),
            datetime(2020, 1, 1, 9, 6, 0),
        ],
        "stock": ["A", "B", "C", "A"],
        "quote": [100, 300, 501, 102],
    }
)

print(df_quotes)
# --8<-- [end:df_quotes]

# --8<-- [start:asof]
df_asof_join = df_trades.join_asof(df_quotes, on="time", by="stock")
print(df_asof_join)
# --8<-- [end:asof]

# --8<-- [start:asof-tolerance]
df_asof_tolerance_join = df_trades.join_asof(
    df_quotes, on="time", by="stock", tolerance="1m"
)
print(df_asof_tolerance_join)
# --8<-- [end:asof-tolerance]

# --8<-- [start:cartesian-product]
tokens = pl.DataFrame({"monopoly_token": ["hat", "shoe", "boat"]})

result = players.select(pl.col("name")).join(tokens, how="cross")
print(result)
# --8<-- [end:cartesian-product]

File: docs/source/src/rust/user-guide/transformations/joins.rs
// --8<-- [start:setup]
use polars::prelude::*;
// --8<-- [end:setup]

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // NOTE: This assumes the data has been downloaded and is available.
    // See the corresponding Python script for the remote location of the data.

    // --8<-- [start:props_groups]
    let props_groups = CsvReadOptions::default()
        .with_has_header(true)
        .try_into_reader_with_file_path(Some("docs/assets/data/monopoly_props_groups.csv".into()))?
        .finish()?
        .head(Some(5));
    println!("{}", props_groups);
    // --8<-- [end:props_groups]

    // --8<-- [start:props_prices]
    let props_prices = CsvReadOptions::default()
        .with_has_header(true)
        .try_into_reader_with_file_path(Some("docs/assets/data/monopoly_props_prices.csv".into()))?
        .finish()?
        .head(Some(5));
    println!("{}", props_prices);
    // --8<-- [end:props_prices]

    // --8<-- [start:equi-join]
    // In Rust, we cannot use the shorthand of specifying a common
    // column name just once.
    let result = props_groups
        .clone()
        .lazy()
        .join(
            props_prices.clone().lazy(),
            [col("property_name")],
            [col("property_name")],
            JoinArgs::default(),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:equi-join]

    // --8<-- [start:props_groups2]
    let props_groups2 = props_groups
        .clone()
        .lazy()
        .with_column(col("property_name").str().to_lowercase())
        .collect()?;
    println!("{}", props_groups2);
    // --8<-- [end:props_groups2]

    // --8<-- [start:props_prices2]
    let props_prices2 = props_prices
        .clone()
        .lazy()
        .select([col("property_name").alias("name"), col("cost")])
        .collect()?;
    println!("{}", props_prices2);
    // --8<-- [end:props_prices2]

    // --8<-- [start:join-key-expression]
    let result = props_groups2
        .clone()
        .lazy()
        .join(
            props_prices2.clone().lazy(),
            [col("property_name")],
            [col("name").str().to_lowercase()],
            JoinArgs::default(),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:join-key-expression]

    // --8<-- [start:inner-join]
    let result = props_groups
        .clone()
        .lazy()
        .join(
            props_prices.clone().lazy(),
            [col("property_name")],
            [col("property_name")],
            JoinArgs::new(JoinType::Inner),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:inner-join]

    // --8<-- [start:left-join]
    let result = props_groups
        .clone()
        .lazy()
        .join(
            props_prices.clone().lazy(),
            [col("property_name")],
            [col("property_name")],
            JoinArgs::new(JoinType::Left),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:left-join]

    // --8<-- [start:right-join]
    let result = props_groups
        .clone()
        .lazy()
        .join(
            props_prices.clone().lazy(),
            [col("property_name")],
            [col("property_name")],
            JoinArgs::new(JoinType::Right),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:right-join]

    // --8<-- [start:left-right-join-equals]
    // `equals_missing` is needed instead of `equals`
    // so that missing values compare as equal.
    let dfs_match = result.equals_missing(
        &props_prices
            .clone()
            .lazy()
            .join(
                props_groups.clone().lazy(),
                [col("property_name")],
                [col("property_name")],
                JoinArgs::new(JoinType::Left),
            )
            .select([
                // Reorder the columns to match the order of `result`.
                col("group"),
                col("property_name"),
                col("cost"),
            ])
            .collect()?,
    );
    println!("{}", dfs_match);
    // --8<-- [end:left-right-join-equals]

    // --8<-- [start:full-join]
    let result = props_groups
        .clone()
        .lazy()
        .join(
            props_prices.clone().lazy(),
            [col("property_name")],
            [col("property_name")],
            JoinArgs::new(JoinType::Full),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:full-join]

    // --8<-- [start:full-join-coalesce]
    let result = props_groups
        .clone()
        .lazy()
        .join(
            props_prices.clone().lazy(),
            [col("property_name")],
            [col("property_name")],
            JoinArgs::new(JoinType::Full).with_coalesce(JoinCoalesce::CoalesceColumns),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:full-join-coalesce]

    // --8<-- [start:semi-join]
    let result = props_groups
        .clone()
        .lazy()
        .join(
            props_prices.clone().lazy(),
            [col("property_name")],
            [col("property_name")],
            JoinArgs::new(JoinType::Semi),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:semi-join]

    // --8<-- [start:anti-join]
    let result = props_groups
        .clone()
        .lazy()
        .join(
            props_prices.clone().lazy(),
            [col("property_name")],
            [col("property_name")],
            JoinArgs::new(JoinType::Anti),
        )
        .collect()?;
    println!("{}", result);
    // --8<-- [end:anti-join]

    // --8<-- [start:players]
    let players = df!(
        "name" => ["Alice", "Bob"],
        "cash" => [78, 135],
    )?;
    println!("{}", players);
    // --8<-- [end:players]

    // --8<-- [start:non-equi]
    let result = players
        .clone()
        .lazy()
        .join_builder()
        .with(props_prices.clone().lazy())
        .join_where(vec![col("cash").cast(DataType::Int64).gt(col("cost"))])
        .collect()?;
    println!("{}", result);
    // --8<-- [end:non-equi]

    // --8<-- [start:df_trades]
    use chrono::prelude::*;

    let df_trades = df!(
        "time" => [
            NaiveDate::from_ymd_opt(2020, 1, 1).unwrap().and_hms_opt(9, 1, 0).unwrap(),
            NaiveDate::from_ymd_opt(2020, 1, 1).unwrap().and_hms_opt(9, 1, 0).unwrap(),
            NaiveDate::from_ymd_opt(2020, 1, 1).unwrap().and_hms_opt(9, 3, 0).unwrap(),
            NaiveDate::from_ymd_opt(2020, 1, 1).unwrap().and_hms_opt(9, 6, 0).unwrap(),
        ],
        "stock" => ["A", "B", "B", "C"],
        "trade" => [101, 299, 301, 500],
    )?;
    println!("{}", df_trades);
    // --8<-- [end:df_trades]

    // --8<-- [start:df_quotes]
    let df_quotes = df!(
        "time" => [
            NaiveDate::from_ymd_opt(2020, 1, 1).unwrap().and_hms_opt(9, 1, 0).unwrap(),
            NaiveDate::from_ymd_opt(2020, 1, 1).unwrap().and_hms_opt(9, 2, 0).unwrap(),
            NaiveDate::from_ymd_opt(2020, 1, 1).unwrap().and_hms_opt(9, 4, 0).unwrap(),
            NaiveDate::from_ymd_opt(2020, 1, 1).unwrap().and_hms_opt(9, 6, 0).unwrap(),
        ],
        "stock" => ["A", "B", "C", "A"],
        "quote" => [100, 300, 501, 102],
    )?;
    println!("{}", df_quotes);
    // --8<-- [end:df_quotes]

    // --8<-- [start:asof]
    let result = df_trades.join_asof_by(
        &df_quotes,
        "time",
        "time",
        ["stock"],
        ["stock"],
        AsofStrategy::Backward,
        None,
        true,
        true,
    )?;
    println!("{}", result);
    // --8<-- [end:asof]

    // --8<-- [start:asof-tolerance]
    let result = df_trades.join_asof_by(
        &df_quotes,
        "time",
        "time",
        ["stock"],
        ["stock"],
        AsofStrategy::Backward,
        Some(AnyValue::Duration(60000, TimeUnit::Milliseconds)),
        true,
        true,
    )?;
    println!("{}", result);
    // --8<-- [end:asof-tolerance]

    // --8<-- [start:cartesian-product]
    let tokens = df!(
        "monopoly_token" => ["hat", "shoe", "boat"],
    )?;

    let result = players
        .clone()
        .lazy()
        .select([col("name")])
        .cross_join(tokens.clone().lazy(), None)
        .collect()?;
    println!("{}", result);
    // --8<-- [end:cartesian-product]

    Ok(())
}
